misc/python/materialize/checks/multiple_partitions.py

```diff
@@ -0,0 +1,103 @@
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+from textwrap import dedent
+from typing import List
+
+from materialize.checks.actions import Testdrive
+from materialize.checks.checks import Check
+
+
+def schemas() -> str:
+    return dedent(
+        """
+       $ set keyschema={
+           "type": "record",
+           "name": "Key",
+           "fields": [
+               {"name": "key1", "type": "string"}
+           ]
+         }
+
+       $ set schema={
+           "type" : "record",
+           "name" : "test",
+           "fields" : [
+               {"name":"f1", "type":"string"}
+           ]
+         }
+    """
+    )
+
+
+class MultiplePartitions(Check):
+    """Test that adds new partitions to a Kafka source"""
+
+    def initialize(self) -> Testdrive:
+        return Testdrive(
+            schemas()
+            + dedent(
+                """
+                $ kafka-create-topic topic=movies-topic
+
+                $ kafka-ingest format=avro key-format=avro topic=movies-topic key-schema=${keyschema} schema=${schema} repeat=1000
+                {"key1": "A${kafka-ingest.iteration}"} {"f1": "A${kafka-ingest.iteration}"}
+
+                > CREATE CONNECTION IF NOT EXISTS kafka_conn FOR KAFKA BROKER '${testdrive.kafka-addr}';
+
+                > CREATE CONNECTION IF NOT EXISTS csr_conn FOR CONFLUENT SCHEMA REGISTRY URL '${testdrive.schema-registry-url}';
+
+                > CREATE SOURCE movies_source
+                  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-movies-topic-${testdrive.seed}', TOPIC METADATA REFRESH INTERVAL MS 500)
+                  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                  ENVELOPE UPSERT
+
+                > CREATE MATERIALIZED VIEW mv_movies AS SELECT * FROM movies_source;
+                
+                $ kafka-add-partitions topic=movies-topic total-partitions=4
+                """
+            )
+        )
+
+    def manipulate(self) -> List[Testdrive]:
+        return [
+            Testdrive(schemas() + dedent(s))
+            for s in [
+                """
+                $ kafka-ingest format=avro key-format=avro topic=movies-topic key-schema=${keyschema} schema=${schema} repeat=40
+                {"key1": "A${kafka-ingest.iteration}"} {"f1": "A${kafka-ingest.iteration}"}
+                """,
+                """
+                $ kafka-ingest format=avro key-format=avro topic=movies-topic key-schema=${keyschema} schema=${schema} repeat=60
+                {"key1": "A${kafka-ingest.iteration}"} {"f1": "A${kafka-ingest.iteration}"}
+                """,
+            ]
+        ]
+
+    def validate(self) -> Testdrive:
+        return Testdrive(
+            dedent(
+                """
+                > SELECT * FROM movies_source_progress;
+                (3,) 0
+                [0,0] 1025
```

Comment:
Would it be more interesting to have a more equal distribution?

Reply:
The distribution when adding new entries is pretty equal. The diff here is because I added lots of entries already at the beginning when only one partition exists. I can play a bit with the numbers, though.